pilot/pkg/bootstrap/trustbundle.go

```diff
@@ -0,0 +1,176 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package bootstrap
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+// source is all possible sources of MeshConfig
+type Source int
+
+type TrustAnchorConfig struct {
+	source Source
+	certs  []string
+}
+
+type TrustAnchorUpdate struct {
+	TrustAnchorConfig
+}
+
+type TrustBundle struct {
+	sourceConfig     map[Source]*TrustAnchorConfig
+	mutex            sync.RWMutex
+	mergedCerts      []string
+	anchorUpdateChan chan *TrustAnchorUpdate
```

Comment:
I am not sure why we need this (or all the associate go routines, periodic waiters, etc). is it possible to just call some `Update(TrustAnchorUpdate)` function on TrustBundle whenever the trust bundle changes?

Reply:
There are multiple go-routines that can change the TrustBundle -> the ones that (eventually) monitor Spiffe endpoints, the callback from MeshConfig Updates, the Callback from CA updates etectera. All of these eventually modify what is essentially a string of certs that need to get pushed to the proxy. Given that the trustbundle is the shared object, we want to prevent these callbacks (some of which may be very time sensitive) to wait on a write lock. I presumed a good strategy would be to serialize any writes through a single goroutine -> This is why all modifiers enqueue their update request and the processUpdates function serializes the updates.

Reply:
it seems simpler to just have the (theoretical) `Update` function add locks?
like

```go
func (tb *TrustBundle) Update() {
tb.mutex.Lock()
defer tb.mutex.Unlock()
...update sourceConfig/mergedCerts...
}
```

Reply:
It would be simpler, but then the goroutine that calls the Update function *could* get locked. For instance, when MeshConfig changes, we add our callback as one of the MeshHandlers. There are other Meshhandlers (a queue of function callbacks), all of which are called within the context of the same goroutine. 
```

	for _, h := range handlers {
		h()
	}
```
Locking here would stall all the other callbacks that need to be invoked. In the current design, the goroutine that processes updates has the explicit job of just updating the trustbundle. Stalling it is acceptable

Reply:
The current design *also* stalls. A write on an unbounded channel is blocking. Also, the time `Update` takes is either O(microsecond) or something is very wrong. Locks are used all of the place, I think we only need to be concerned about this issue if the work being done is something that may not terminate in a reasonable time.

Reply:
Edit: Sorry, there was code I had not pushed that increased the length of the update channel so that writers writing into the channel would not block.

Reply:
While the Update currently may not take much time, if we increase the size of the callback (to do verification and other activities) it could affect the time taken in the callback - definitely not longer than microseconds, but still. Its worth doing all this within a seperate goroutine IMO. Is the added complexity not worth it in your opinion?

Reply:
It is not worth it at all in my opinion. A callback pattern is very standard both in Istio and other code base (javascript :slightly_smiling_face: ). We use it in far more critical code paths (and slower) and have never had issues with this.

This code will be invoked probably 1x at startup and never again, we don't need to prematurely optimize this

Reply:
Callbacks are fairly standard. However, I have spent a lot of time debugging issues in the past where a timer callback took too much time and delayed all the other callbacks in the same bucket. The ability to offload the time-sensitive part of the callback to a goroutine is a key advantage of golang. Perhaps there is some hidden cost to using the goroutine here that I am not aware of? Either way, will remove it for now.

Reply:
If the callback is only on the root rotation work, can we just remove that callback because it's a corner case?
If the merge function is the contention part, I don't think it's going to be costly :)
Shankar I understand you want to write high-quality code but the queuing thing is harder to maintain than a simple lock.

Reply:
Sure. Removed update goroutine and expanding scope of locks